### PR TITLE
Module 4

### PR DESCRIPTION
**Pull request by TheLegendaryPan**
Combined fixes to Module 3 into Module 4 project branch. 
Module 3 changes include updated selenium board creation and board_id reference. Now test board gets deleted. 
Added selenium new test case to create card and validate if new card exists. 

Module 4 changes include both Vagrant and Poetry VM changes. VM created successfully and app launched as per Vagrant up command post poetry installation on VM. 

Please review.

### CODE REVIEW

File: Vagrantfile [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  # SHELL
+	config.vm.network "forwarded_port", guest: 5000, host: 5000
+	config.vm.provision "shell", privileged: false, inline: <<-SHELL     
+		sudo apt-get update 
+		#install pre-req
+		sudo apt-get install -y build-essential libssl-dev zlib1g-dev libbz2-dev \
+		libreadline-dev libsqlite3-dev wget curl llvm libncurses5-dev libncursesw5-dev \
+		xz-utils tk-dev libffi-dev liblzma-dev python-openssl git
+		
+		#apt-get install pyenv as vagrant, not root
```
Comment (hugh-emerson) [inline Vagrantfile]: A lot of commented lines in this file can be deleted. This is an example of one that looks out of date and is somewhat misleading.

File: Vagrantfile [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+		git clone https://github.com/pyenv/pyenv.git ~/.pyenv
+		echo 'export PYENV_ROOT="$HOME/.pyenv"' >> ~/.bash_profile
+		echo 'export PATH="$PYENV_ROOT/bin:$PATH"' >> ~/.bash_profile
+		echo -e 'if command -v pyenv 1>/dev/null 2>&1; then\n  eval "$(pyenv init -)"\nfi' >> ~/.bash_profile
+		
+		#install pip then python, otherwise pyenv install hangs
+		sudo apt-get -y install python-pip
```
Comment (hugh-emerson) [inline Vagrantfile:18]: Why do you need this?

File: Vagrantfile [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	config.trigger.after :up do |trigger|     
+		trigger.name = "Launching App"     
+		trigger.info = "Running the TODO app setup script"     
+		trigger.run_remote = {privileged: false, inline: "
+		# Install dependencies and launch
+		cd /vagrant
+		poetry update
+		poetry install
+		# <your script here> 
+		nohup poetry run flask run --host=0.0.0.0 ##just nohup works but doesnt return prompt
```
Comment (hugh-emerson) [inline Vagrantfile]: Try `nohup poetry run flask run --host=0.0.0.0 > log.txt 2>&1 &`

The `> log.txt 2>&1` is to redirect the stdout output to a file called log.txt, and to do the same with stderr.

The `&` at the end is to run the command in a separate process, often paired with `nohup ... &` so it's a standalone process and not a child.

**Review by hugh-emerson — APPROVED** [review]
Good job! Just some minor comments

Somewhat importantly though, you should update the README.
You've managed to switch to poetry but the README needs to match that:

It should tell the reader to copy .env.template to .env and to run `poetry install` / `poetry run flask run` / `poetry run pytest`  
And also that the reader can run `vagrant up` if they have Vagrant and a hypervisor. 

You can also delete the setup.sh file now.

File: Vagrantfile (hunk 3 above)
Comment (TheLegendaryPan) [reply Vagrantfile]: That worked thanks! i can see the log also capturing the connection requests when I navigates the pages. Thank you!

**Pull request merged**